knimin/lib/data_access.py

```diff
@@ -3165,96 +3031,128 @@ def read_sample_plate(self, plate_id):
+
+            # We can set the new layout. First clear the previous layout
+            # and insert the new one
+            self._clear_sample_plate_layout(sample_plate_id)
+            sql = """INSERT INTO pm.sample_plate_layout
+                        (sample_plate_id, sample_id, row, col, name, notes)
                      VALUES (%s, %s, %s, %s, %s, %s)"""
-            for x in layout:
-                sample_id = x['sample_id']
-                if not self._sample_exists(sample_id):
-                    raise ValueError('Sample ID %s does not exist.'
-                                     % sample_id)
-                TRN.add(sql, [id, sample_id, x['col'], x['row'],
-                              x.get('name'), x.get('notes')])
-                TRN.execute()
+            sql_args = []
+            for row_idx, row in enumerate(layout):
+                for col_idx, values in enumerate(row):
+                    sql_args.append([sample_plate_id, values['sample_id'],
+                                     row_idx, col_idx, values['name'],
+                                     values['notes']])
+            TRN.add(sql, sql_args, many=True)
+            TRN.execute()
 
-    def read_sample_plate_layout(self, id):
+    def read_sample_plate_layout(self, sample_plate_id):
         """Reads the layout of a sample plate
 
         Parameters
         ----------
-        id : int
+        sample_plate_id : int
             ID of the sample plate whose layout is to be read
 
         Returns
         -------
-        list of dict
-            {sample_id : str, col : int, row : int, name : str, notes : str}
-            A list of sample-to-well records, each of which includes:
-            Sample ID, column number, row number, name, and notes
-            The list is sorted by column then by row in ascending order
+        layout : list of list of dict
+            A 2-D matrix storing the per well information in a dict with the
+            format: {'sample_id': str, 'name': str, 'notes': str}
         """
         with TRN:
-            self._sample_plate_exists(id)
-            if not self._sample_plate_layout_exists(id):
-                return []
+            self._sample_plate_exists(sample_plate_id)
             sql = """SELECT sample_id, col, row, name, notes
                      FROM pm.sample_plate_layout
                      WHERE sample_plate_id = %s
-                     ORDER BY col, row"""
-            TRN.add(sql, [id])
-            return [dict(x) for x in TRN.execute_fetchindex()]
+                     ORDER BY row, col"""
+            TRN.add(sql, [sample_plate_id])
+            res = TRN.execute_fetchindex()
+            layout = []
+            row = []
+            # To construct the output layout we just need to iterate over
+            # the results since they are correctly sorted
+            for well in res:
+                # Transform the DictCursos to an actual dictionary
```

Comment:
Is this DictCursors? 

I didn't know about DictCursors so I was wondering what a DictCurso was. :D #spanishpython